File: branches/scala_implementation/taboosearch/src/taboosearch/Selector.scala
package taboosearch
import scala.annotation.unchecked.uncheckedVariance
import core._

abstract class Selector[+S <: Solution] extends core.Selector[Generation[S] @uncheckedVariance] with GenerationBuilder[S] {
	
	override def keepTheBestSolutions(generation : Generation[S] @uncheckedVariance,
							 ctx : Context[Solution] @uncheckedVariance) : Generation[S] = {
		
		var taboolator = ctx.taboolator;
		var evaluator = ctx.evaluator;
		
		var candidateFitness = Double.MaxValue;
		var candidate : Option[S] = None;

		for (solution <- generation) {
			if (!taboolator.isTabu(solution)) {
				if (evaluator.evaluate(solution) < candidateFitness) {
					candidateFitness = evaluator.evaluate(solution);
					candidate = Some(solution);
				}
			}
		}

		val result = createGeneration(); // method from GenerationBuilder trait
		candidate match {
			case Some(solution) => {
				taboolator.setTabu(solution);
				result += solution;
			}
			case None => println("Please decrease taboo tenure!")
		}
		return result;
	}

}

File: src/diffevolution/Selector.java
package diffevolution;

import diffevolution.Context;
import diffevolution.Generation;
import diffevolution.Solution;

public class Selector < S extends ArraySolution<?>,
						G extends Generation<S>, 
						Cr extends AbstractCrossoverFabric<S>,
						C extends Context<G, S, Cr>>
						implements core.Selector<G> {
	
	private C context;
	
	public Selector(C context) {
		this.context = context;
	}

	@Override
	public G keepTheBestSolutions(G g, G currentGeneration) {
		for (int i = 0; i < g.size(); ++i) {
			if (context.getEvaluator().evaluate(g.get(i)) < context.getEvaluator().evaluate(currentGeneration.get(i))) {
				currentGeneration.set(i, g.get(i));
			}			
		}
		return currentGeneration;
	}

}
